aws_lambda_powertools/utilities/feature_toggles/configuration_store.py

```diff
@@ -0,0 +1,207 @@
+        if region_name is not None:
+            config = Config(region_name=region_name)
+        self._conf_store = AppConfigProvider(environment=environment, application=service, config=config)
+
+    def _validate_json_schema(self, schema: str) -> bool:
+        ## todo
+        return True
+
+    def _match_by_action(self, action: str, restriction_value: Any, context_value: Any) -> bool:
+        mapping_by_action = {
+            ACTION.EQUALS.value: lambda a, b: a == b,
+            ACTION.STARTSWITH.value: lambda a, b: a.startswith(b),
+            ACTION.ENDSWITH.value: lambda a, b: a.endswith(b),
+            ACTION.CONTAINS.value: lambda a, b: a in b,
+        }
+
+        try:
+            func = mapping_by_action.get(action, lambda a, b: False)
+            return func(context_value, restriction_value)
+        except Exception:
+            return False
+
+    def _handle_rules(
+        self,
+        logger: object,
+        feature_name: str,
+        rules_context: Dict[str, Any],
+        default_value: bool,
+        rules: List[Dict[str, Any]],
+    ) -> bool:
+        for rule in rules:
+            rule_name = rule.get(RULE_NAME_KEY, "")
+            rule_default_value = rule.get(RULE_DEFAULT_VALUE)
+            is_match = True
+            restrictions: Dict[str, str] = rule.get(RESTRICTIONS_KEY)
+            if restrictions is None:
+                error_str = f"invalid rule schema detected in rule {rule_name}, missing restrictions list"
+                logger.error(error_str)
+                raise ConfigurationException(error_str)
+
+            for restriction in restrictions:
+                context_value = rules_context.get(restriction.get(RESTRICTION_KEY, ""))
+                if not context_value:
+                    logger.debug(
+                        "rule did not not match key",
+                        rule_name=rule_name,
+                        default_value=rule_default_value,
+                        feature_name=feature_name,
+                    )
+                    is_match = False  # rule doesn't match, continue to next one
+                    break
+                elif not self._match_by_action(
+                    restriction.get(RESTRICTION_ACTION), restriction.get(RESTRICTION_VALUE), context_value
+                ):
+                    logger.debug(
+                        "rule did not match action",
+                        rule_name=rule_name,
+                        default_value=rule_default_value,
+                        feature_name=feature_name,
+                    )
+                    is_match = False  # rules doesn't match restriction
+                    break
+            # if we got here, all restrictions match
+            if is_match:
+                logger.debug(
+                    "rule matched", rule_name=rule_name, default_value=rule_default_value, feature_name=feature_name
+                )
+                return rule_default_value
```

Comment:
Not sure I understand the difference between rule_default_value and default_value. Is rule_default_value similar to value_if_true?

Reply:
there are 3 default values.
1. the default value you provide in the get_feature_toggle api,. This is the value that will be returned in case the feature doesnt appear in the schema (someone broke the schema, removed it by mistake etc.).
2. each feature has a default value. However, it can be overridden by a rule's default value in case of a rule match.
3. rule's default value - naming is not great i'd admit. In case of a rule match, this is the toggle's returned value, it can be different that the value of the feature toggle we defined in section 2.

Reply:
Yes, naming is one of the hardest pieces of programming :) - having three "default" values will be confusing. Also, not sure about the last one, as I can see how returning the value of the toggle itself may be useful and simpler (the default behaviour value can be changed on the rule without having to change the code) - thinking about naming: value_if_missing, feature_default_value ?

Reply:
what about the third one? the rule value which overrides the feature_default_value. rule_default_value?
I'll change it to rule_default_value for now

Reply:
Yes, rule_default_value sounds ok but my only question here is why do we want to return a default value and not the toggle value directly. If all the previous checks have matched, the function should return the value of the toggle directly. Can't see the benefit of a default value here.

Reply:
That's the magic of the rule engine and the most important part. You want to have the toggle False (default_feature_toggle) for most use case. However, if a rule matches (rule match is the same as a feature exists) then you might want to return a different value.
This is useful when you want to have a feature turned off for most customers but turned on for customer X. So yuo define a rule that will match customer X and only customer X and have that rule with rule_default_value set to True.

Reply:
I'm a bit confused now (my bad) so let me try to recap the behaviour as much as I understand it:
* if the rule matches, it returns rule_default_value 
* if it doesn't or is missing, it returns feature_default_value 
* only if the feature doesn't exist you return value_if_missing
This allows for only two possible responses: rule_default_value or feature_default_value

In this case, I guess rule_default_value should be better named value_when_applies to show that this value will be returned only when all restrictions are matched.

Reply:
yes you are correct. i've renamed it and pushed the fixes